keycloak/docker/src/main/docker/keycloak/mitigate-cves.java

```diff
@@ -0,0 +1,478 @@
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.security.MessageDigest;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+
+class MitigateCves {
+
+    static final boolean VERIFY_CHECKSUM = Boolean.getBoolean("verifyChecksum");
+
+    static final boolean FAIL_ON_FAILED_MITIGATION = Boolean.getBoolean("failOnFailedMitigation");
+
+    public static void main(String[] args) {
+
+        System.out.println("### Begin CVE mitigation...");
+
+        String keycloakHomeFolder = Optional.ofNullable(System.getenv("KEYCLOAK_HOME")).orElse("/opt/jboss/keycloak");
+        String baseModulesFolder = keycloakHomeFolder + "/modules/system/layers/base";
+
+        List<CveFix> cveFixes = Arrays.asList(
+
+                // Remove unused hornetq modules with CVEs
+                new CveFixDeleteLibrary(baseModulesFolder + "/org/hornetq"),
+
+                // Remove unused activemq modules with CVEs
+                new CveFixDeleteLibrary(baseModulesFolder + "/org/apache/activemq"),
+                new CveFixDeleteLibrary(baseModulesFolder + "/org/wildfly/extension/messaging-activemq"),
+                new CveFixReplaceInFile(baseModulesFolder + "/org/jboss/jts/main/module.xml"
+                        , "<module name=\"org.apache.activemq.artemis.journal\"/>"
+                        , ""
+                ),
+
+                // Remove unused jboss-client.jar library with CVEs
+                new CveFixDeleteLibrary(keycloakHomeFolder + "/bin/client/jboss-client.jar"),
+
+                // Replace module jars with CVEs with the latest compatible fixed version.
```

Comment:
Is Keycloak shipping with vulnerable dependencies? Was trying to relate the library upgrade to a specific CVE but was unsuccessful. How did you determine these binaries needed to be upgraded?

Reply:
Yes as many other large open-source projects Keycloak ships some libraries that have known CVEs that might be exploited by adversaries.

We use aquasec/trivy to look for jar's with CVEs.
https://github.com/thomasdarimont/keycloak-project-example#scan-image-for-vulnerabilities

All mentioned fixes are related to CVE's reported by trivy. I'm going to annotate the addressed CVEs asap.